Added function accurateCastOrNull, allow different types inside IN subquery

kitaisreal opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Mentioned in https://github.com/ClickHouse/ClickHouse/issues/10290.
Mentioned in https://github.com/ClickHouse/ClickHouse/issues/10266.

Changelog category (leave one):
- New Feature


Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added function `accurateCastOrNull`. This closes #10290. Add type conversions in `x IN (subquery)` expressions. This closes #10266.

kitaisreal committed:
Added function accurate cast or null

kitaisreal commented on src/Functions/FunctionsConversion.h:
```diff
@@ -147,22 +177,53 @@ struct ConvertImpl
+                    if constexpr (convert_strategy == ConvertStrategy::Accurate)
+                    {
+                        /// TODO: Add support for big integers
+                        if (accurate::greaterOp(std::numeric_limits<ToFieldType>::max(), vec_from[i]) ||
```
@alexey-milovidov not sure how to properly implement it for types without numeric limits, do we have some places where similar functionality implemented ? Thanks.

alexey-milovidov replied:
The most simple is to implement numeric limits for bigints.

kitaisreal committed:
Added numeric limits for UInt128

kitaisreal committed:
Fixed setting null value in vector. Updated tests.

kitaisreal committed:
Updated documentation

kitaisreal committed:
Added Decimal support

kitaisreal committed:
Fixed style issues

kitaisreal committed:
Fixed clang compile issues

kitaisreal commented on src/Core/Settings.cpp:
```diff
@@ -16,9 +16,12 @@ namespace ErrorCodes
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wframe-larger-than="
 
 IMPLEMENT_SETTINGS_TRAITS(SettingsTraits, LIST_OF_SETTINGS)
```
@alexey-milovidov latest clang warning
```
Documents/ClickhouseBrowse/src/Core/Settings.cpp:20:1: error: stack frame size of 81128 bytes in lambda expression [-Werror,-Wframe-larger-than=]
IMPLEMENT_SETTINGS_TRAITS(SettingsTraits, LIST_OF_SETTINGS)
^
Documents/ClickhouseBrowse/src/Core/BaseSettings.h:817:46: note: expanded from macro 'IMPLEMENT_SETTINGS_TRAITS'
        static const Accessor the_instance = [] \
```
```
Documents/ClickhouseBrowse/build$ clang++-11 --version
Ubuntu clang version 11.0.1-++20201103062930+ef4ffcafbb2-1~exp1~20201103053545.129
```

alexey-milovidov replied:
Ok. Frame size limit is currently just a precaution.

kitaisreal committed:
Fixed compile issues

kitaisreal committed:
Added const to arguments

kitaisreal committed:
Fixed compile issues

kitaisreal committed:
Allow different types inside IN subquery

kitaisreal commented:
@alexey-milovidov make test implementation of allowing different types inside in subquery. Please check 01600_select_in_different_types.sql and Set.cpp for a reference.

alexey-milovidov commented on tests/queries/0_stateless/01600_select_in_different_types.sql:
```diff
@@ -0,0 +1,30 @@
+
+SELECT 1 IN (1);
+-- Right now this working because of accurate cast. Need to discuss.
+SELECT '1' IN (SELECT 1);
```
It's ok to keep.

alexey-milovidov replied:
It's more aligned with the behaviour of standard SQL where weak typing is used.
We don't like weak typing :) But we tend to make our SQL more compatible with standard (through some settings or without). If this behaviour is "for free", let's keep it without introducing any settings.

kitaisreal replied:
@alexey-milovidov It is for free for cases where we insert data in set from subselect, but when we make set from collection during AST visiting, we use convertFieldToType and it will throw exception if types does not match.

Example: SELECT '1' IN (1); will throw exception.

It does not make a lot of changes to update this behaviour. Should we fix it ?

alexey-milovidov replied:
It looks inconsistent but let's keep it as is for now.

alexey-milovidov commented on src/Core/AccurateComparison.h:
```diff
@@ -515,15 +515,24 @@ inline bool NO_SANITIZE_UNDEFINED convertNumeric(From value, To & result)
+        }
+    }
+
+    if (accurate::greaterOp(value, std::numeric_limits<To>::max())
```
What if we convert Float64's inf to Float32's inf or vice versa?

alexey-milovidov replied:
Also https://en.cppreference.com/w/cpp/types/numeric_limits/lowest ?

kitaisreal replied:
Will investigate floating point types conversion, seems like there is are a bunch of problems.

kitaisreal replied:
Related to infinite conversions we can add special case above in addition to isNan.

alexey-milovidov replied:
Ok.

alexey-milovidov replied:
BTW, loss of precision is Ok. For example,

`toFloat32(0.1) IN (SELECT 0.1)` should work.

alexey-milovidov replied:
So, we can assume all the range of [lowest, max] of Float32 in Float64 to be representable in Float32.

alexey-milovidov replied:
(actually I'm not 100% sure about it)

kitaisreal replied:
Updated. As discussed does not create specialization for case like ```toFloat32(0.1) IN (SELECT 0.1)```.

alexey-milovidov commented on src/Core/AccurateComparison.h:
```diff
@@ -515,15 +515,24 @@ inline bool NO_SANITIZE_UNDEFINED convertNumeric(From value, To & result)
 
     result = static_cast<To>(value);
-    return equalsOp(value, result);
+    return true;
```
Also what's about the case of conersion 1.5 to integer type (should return false)?

PS. Exact equality is Ok, so 1.000...1 also not equals to 1.

kitaisreal replied:
Agreed equalsOp is required.

kitaisreal committed:
Added equality comparsion

alexey-milovidov commented on src/Interpreters/Set.cpp:
```diff
@@ -254,8 +257,15 @@ ColumnPtr Set::execute(const Block & block, bool negative) const
-        checkTypesEqual(i, block.safeGetByPosition(i).type);
-        materialized_columns.emplace_back(block.safeGetByPosition(i).column->convertToFullColumnIfConst());
+        /// TODO: Optimize making cast only if types different
+        /// TODO: This case SELECT '1' IN (SELECT 1); should not work but with AccurateCastOrNull it works
```
This is Ok :)

alexey-milovidov replied:
BTW we also have obscure setting `transform_null_in`.

We can disable conversions (and resort to strict type checking) when it's enabled.

kitaisreal replied:
Updated as we discussed, adding additional function accurateCast.

alexey-milovidov commented on src/Interpreters/Set.cpp:
```diff
@@ -30,6 +32,7 @@
 #include <ext/range.h>
 #include <DataTypes/DataTypeLowCardinality.h>
 
+#include <iostream>
```
Remove before merge...

kitaisreal replied:
Fixed.

kitaisreal committed:
Updated usage of different types during IN query

1. Added accurateCast function.
2. Use accurateCast in Set during execute.
3. Added accurateCast tests.
4. Updated select_in_different_types tests.

kitaisreal committed:
Fixed compile issues

kitaisreal committed:
Fixed Set execute for columns that cannot be inside Nullable

kitaisreal committed:
Fixed clang build issue

alexey-milovidov committed:
Update castColumn.h

alexey-milovidov committed:
Update castColumn.h

kitaisreal committed:
Updated non-finite floating conversion into int

kitaisreal committed:
Fixed tests

kitaisreal committed:
Added tryParseDecimal functions

kitaisreal committed:
Fixed tests

alexey-milovidov committed:
Update DecimalFunctions.h

kitaisreal committed:
Fixed compilation timeout

kitaisreal committed:
Fixed ya.make

alexey-milovidov commented:
@kitaisreal This is somehow related: https://clickhouse-test-reports.s3.yandex.net/19677/3b6395fbb774c132273083f4f84a4f5a34c12866/fuzzer_debug/server.log
